### ISSUE TITLE
FR: Reading multiple CSVs from a single file

### ISSUE THREAD

**Issue by charsbar** [Tux/Text-CSV_XS#41]
re: https://github.com/makamaka/Text-CSV/issues/62

> Occasionally I have to process files that contain multiple CSVs. Each of these CSVs is stored in the file as a heading, followed by the data lines, followed by an empty line (or eof).
> 
> It would be nice if Text::CSV had an option that basically says: stop reading after an empty line. This would make it possible to write something similar to:
> 
> csv (in => $fh,  out => \@aoh1, stop_at_empty => 1);
> csv (in => $fh,  out => \@aoh2, stop_at_empty => 1);
> csv (in => $fh,  out => \@aoh3, stop_at_empty => 1);

**Comment by sciurius** [comment]
1. `skip_empty_rows => "stop"` is not understood:

````
Argument "stop" isn't numeric in subroutine entry at /usr/lib64/perl5/vendor_perl/Text/CSV_XS.pm line 1445,
````

2. It seems that with `skip_empty_rows => "stop"` it reads past the empty line and consumes the next line (which is the heading for the next CSV segment).

````
% perl test.pl test.zip
[
  {
    _id => 'Group1',
  },
  {
    _id => 'Group2',
  },
]
POS: 140
# CSV_XS ERROR: 1012 - INI - the header contains an empty field @ rec 1 pos 0
INI - the header contains an empty field at test.pl line 20.
Exit 255
````
POS should be 19 after reading the first CSV segment.



[test.zip](https://github.com/Tux/Text-CSV_XS/files/11145467/test.zip)

**Comment by sciurius** [comment]
This is a full test case for the project I'm working on.
Input is a member of a zip file, read with Archive::Zip::MemberRead. Line terminators are CRLF.

[test.zip](https://github.com/Tux/Text-CSV_XS/files/11147130/test.zip)

**Comment by Tux** [comment]
```
my $ifn = shift or die "usage: $0 file.csv\n";

my %dta;
my $sct = "pt0";
my @hdr;

csv (in    => $ifn,
     out   => undef,
     on_in => sub {
        if (@{$_[1]} == 1 && $_[1][0] eq "") {  # Empty row
            @hdr = ();
            $sct++;
            return;
            }
        unless (@hdr) {
            @hdr = @{$_[1]};
            return;
            }
        my %r; @r{@hdr} = @{$_[1]};
        push @{$dta{$sct}} => \%r;
        },
    );
```
Will be the current supported approach for your test-case

I'll look into the `provider` suggestion, which I like

**Comment by Tux** [comment]
FWIW `skip_empty_rows` now supports `0`, `1`, `die`, `croak`, `stop`, `eof` en callbacks. Commits have been pushed https://github.com/Tux/Text-CSV_XS/commit/1e90c32e1a80f23406f3db9c2bc03a2507dd7c72

**Comment by Tux** [comment]
Closed now that 1.51 is out in the wild. I'll bear `provider` in mind
